source/Agents/PathAgent.cpp

```diff
@@ -0,0 +1,231 @@
+/**
+ * @file PathAgent.cpp
+ * @author David Rackerby
+ */
+
+#include "PathAgent.hpp"
+
+#include <regex>
+#include <stdexcept>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "../core/AgentBase.hpp"
+#include "../core/GridPosition.hpp"
+
+namespace walle {
+
+/**
+ * Constructor (vector)
+ * @param id unique agent id
+ * @param name name of path agent
+ * @param offsets collection of offsets to move the agent
+ * @attention The sequence of offsets must not be empty
+ */
+PathAgent::PathAgent(size_t id, std::string const& name,
+                     std::vector<cse491::GridPosition> && offsets) : cse491::AgentBase(id, name), offsets_(offsets) {
+  assert(!offsets_.empty());
+}
+
+/**
+ * Constructor (string)
+ * @param id unique agent id
+ * @param name name of path agent
+ * @param commands sequence of commands to be interpreted as offsets
+ * @attention The sequence of offsets must not be empty
+ */
+PathAgent::PathAgent(size_t id, std::string const& name,
+                     std::string_view commands) : cse491::AgentBase(id, name), offsets_(str_to_offsets(commands)) {
+  assert(!offsets_.empty());
+}
+
+/**
+ * Checks that the agent is able to freely change its own grid location
+ * Verifies that it can currently index into a valid offset
+ * @return true if so; false otherwise
+ */
+bool PathAgent::Initialize() {
+  return HasAction("move_self") && index_ >= 0 && static_cast<size_t>(index_) < offsets_.size();
+}
+
+/**
+ * Updates the agent's position by applying the current offset
+ */
+void PathAgent::Update() {
+  cse491::GridPosition pos_offset = offsets_[index_];
+  SetPosition(position.GetX() + pos_offset.GetX(), position.GetY() + pos_offset.GetY());
```

Comment:
We are not supposed to move the agents position directly. We need to tell the world which direction we are moving and then the world will move us.

Reply:
The class has been refactored to instead provide the agent's next position to allow outside code to change the position rather than that behavior being restricted solely to the PathAgent class